tools/codegen/gen.py

```diff
@@ -243,13 +248,111 @@ def __call__(self, f: Union[StructuredNativeFunctions, NativeFunction]) -> List[
         else:
             assert_never(f)
 
+    def gen_structured_class_set_output(self, k: SchemaKind, parent_class: str, generate_super: bool) -> str:
+        if generate_super:
+            set_output_super = f"{parent_class}::set_output(output_idx, sizes, strides, options, names);"
+        else:
+            set_output_super = ""
+        return f"""
+void set_output(int64_t output_idx, IntArrayRef sizes, IntArrayRef strides,
+                TensorOptions options, DimnameList names) override {{
+    {self.gen_structured_class_set_output_body(k)}
+    if (!names.empty()) namedinference::propagate_names(outputs_[output_idx], names);
+    // super must happen after, so that downstream can use maybe_get_output
+    // to retrieve the output
+    {set_output_super}
+}}
+"""
+
+    def gen_structured_class_set_output_body(self, k: SchemaKind) -> str:
+        if k is SchemaKind.functional:
+            if self.dispatch_key == "Meta":
+                return """
+if (strides.empty()) {
+    outputs_[output_idx] = at::empty_meta(sizes, options);
+} else {
+    TORCH_INTERNAL_ASSERT(0, "not implemented yet");
+}
+"""
+            else:
+                expanded_topts = "optTypeMetaToScalarType(options.dtype_opt()), options.layout_opt(), " \
+                    "options.device_opt(), options.pinned_memory_opt()"
+                if self.dispatch_key == "CPU":
+                    empty_impl = "at::native::empty_cpu"
+                    empty_strided_impl = "at::native::empty_strided_cpu"
+                elif self.dispatch_key == "CUDA":
+                    empty_impl = "at::native::empty_cuda"
+                    empty_strided_impl = "at::native::empty_strided_cuda"
+                else:
+                    raise AssertionError("unsupported dispatch key")
+                return f"""
+if (strides.empty()) {{
+    outputs_[output_idx] = {empty_impl}(sizes, {expanded_topts}, options.memory_format_opt());
+}} else {{
+    outputs_[output_idx] = {empty_strided_impl}(sizes, strides, {expanded_topts});
+}}
+"""
+        elif k is SchemaKind.inplace:
+            return "// TODO: consistency check?"
+        elif k is SchemaKind.out:
+            return """
+at::native::resize_output(outputs_[output_idx], sizes);
+if (!strides.empty()) {
+    TORCH_INTERNAL_ASSERT(!options.memory_format_opt().has_value());
+    at::native::as_strided_(outputs_[output_idx], sizes, strides);
```

Comment:
can it ever happen that output was correctly sized, but strides that were passed here are not the same as output strides? Because in this case output would be restrided unexpectedly.

Reply:
Good question. I don't think TensorIterator will ever do this, but I do I think we are providing an error prone API. I'll test some tweaks to it, where we NEVER restride if the tensor was correctly sized.

This poses an interesting question of what we should do if (1) the operator advertises that strides are unspecified on set_output, and (2) the operator advertises that strides are contiguous in set_output. Let's consider (2) first. If a user says that the function always returns a contiguous tensor, then it seems safe to assume that it isn't going to work correctly if it is given a non-contiguous out tensor. But there is something pretty nifty we could do here: we could allocate (and pass in) a temporary contiguous out tensor, and then copy it to the noncontiguous out tensor at the end. This is fairly analogous to the dtype temporaries that TensorIterator does when the out tensor doesn't have the correct computation dtype.

What about if stride is not specified? I feel like this should implicitly mean that contiguous is being requested. So then the same behavior in (2) would also apply.

Reply:
(2) happens in a lot of manual implementations - if user-passed `out` does not agree with what our libraries/implementations can handle, we use temporaries and then copy then to `out`.

Reply:
There is also the situation where `resize_outputs(false)` in the TensorIterator config. In that case, the output also participates in `infer_size` broadcasting, and if it it is incompatible TensorIterator will raise an error. So this sounds like we also have a latent bug here:

* If `resize_outputs(false)`, and
* Output shape is not the same as `shape_` (e.g., if broadcasting into output shape happens)

we will attempt to resize output in this case anyway (ouch). This doesn't apply to add because add is OK with resizing.

Probably the easiest way to fix this is to check if `will_resize` before calling `set_output`; if `will_resize` is false, we unconditionally feed in the *original* shape of the output tensor. One thing that kind of sucks here is `set_output` will dutifully do another shape comparison to confirm that no resizing is necessary, but this really isn't necessary (I suppose you could add another method for specifically this situation).